geth/peers/peerpool.go

```diff
@@ -222,37 +229,65 @@ func (p *PeerPool) handleServerPeers(server *p2p.Server, events <-chan *p2p.Peer
 				retryDiscv5 = time.After(discoveryRestartTimeout)
 				log.Error("starting discv5 failed", "error", err, "retry", discoveryRestartTimeout)
 			}
+		case <-stopDiscv5:
+			p.handleStopTopics(server)
 		case event := <-events:
 			switch event.Type {
 			case p2p.PeerEventTypeDrop:
 				log.Debug("confirm peer dropped", "ID", event.Peer)
 				if p.handleDroppedPeer(server, event.Peer) {
 					retryDiscv5 = time.After(0)
 				}
-			case p2p.PeerEventTypeAdd:
+			case p2p.PeerEventTypeAdd: // skip other events
 				log.Debug("confirm peer added", "ID", event.Peer)
-				if p.opts.AllowStop && p.handleAddedPeer(server, event.Peer) {
-					log.Debug("closing discv5 connection because all topics reached max limit", "server", server.Self())
-					p.stopDiscovery(server)
-				}
-			default: // skip other events
+				p.handleAddedPeer(server, event.Peer)
+				stopDiscv5 = time.After(p.opts.TopicStopSearchDelay)
+			default:
 				continue
 			}
 			SendDiscoverySummary(server.PeersInfo())
 		}
 	}
 }
 
-// handleAddedPeer notifies all topics about added peer and return true if all topics has max limit of connections
-func (p *PeerPool) handleAddedPeer(server *p2p.Server, nodeID discover.NodeID) (all bool) {
+// handleAddedPeer notifies all topics about added peer.
+func (p *PeerPool) handleAddedPeer(server *p2p.Server, nodeID discover.NodeID) {
 	p.mu.Lock()
 	defer p.mu.Unlock()
-	all = true
 	for _, t := range p.topics {
 		t.ConfirmAdded(server, nodeID)
 		if p.opts.AllowStop && t.MaxReached() {
+			t.setStopSearchTimeout(p.opts.TopicStopSearchDelay)
+		}
+	}
+}
+
+// handleStopTopics stops the search on any topics having reached its max cached
+// limit or its delay stop is expired, additionally will stop discovery if all
+// peers are stopped.
+func (p *PeerPool) handleStopTopics(server *p2p.Server) {
+	if !p.opts.AllowStop {
+		return
+	}
+	for _, t := range p.topics {
+		if t.readyToStopSearch() {
 			t.StopSearch()
-		} else {
+		}
+	}
+	if p.allTopicsStopped() {
+		log.Debug("closing discv5 connection because all topics reached max limit", "server", server.Self())
+		p.stopDiscovery(server)
+	}
+}
+
+// allTopicsStopped return true if all topics are stopped
+func (p *PeerPool) allTopicsStopped() (all bool) {
+	if !p.opts.AllowStop {
```

Comment:
Hmm, this usage shows that a naming like `AllowsStop` would match better. But not important.